feat(platform-browser): Expose EventManagerPlugin in the public api

JeanMeche opened the pull request:
This commit exposes the `EventManagerPlugin` abstract class to the public API.
It provides a documentation with a usecase and a basic implementation describing the possibilities

Rationale for opening to the public API :

1. The `EVENT_MANAGER_PLUGINS` token is already public
2. The documentation provides a usefull usecase
3. Plugins can already be implemented by following the same API
4. This does not increase the API surface because of 3.
5. This topic gained some visibility in the last few days in the twitter community and seems to be not so widely known, so this helps also for the visibility ! (Also wanted to share the usescases implemented by [@tinkoff/ng-event-plugins](https://github.com/Tinkoff/ng-event-plugins))

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->


- [x] Feature

## Does this PR introduce a breaking change?


- [x] No

pweyrich commented on packages/platform-browser/src/dom/events/event_manager.ts:
```diff
@@ -9,7 +9,7 @@
 
 /**
- * The injection token for the event-manager plug-in service.
+ * The injection token for the for plugins of the `EventManager` service.
```
```suggestion
 * The injection token for plugins of the `EventManager` service.
```

JeanMeche commented:
Can I get the [aio: preview](https://github.com/angular/angular/pulls?q=is%3Aopen+is%3Apr+label%3A%22aio%3A+preview%22) tag ?

waterplea commented:
Super glad to see movement in this direction 👍

Just wanted to mention these two issues slightly impeding wide adoption of custom plugins:

https://github.com/angular/angular/issues/37850
https://github.com/angular/angular/issues/40553

JeanMeche commented on goldens/public-api/platform-browser/index.md:
```diff
@@ -92,6 +92,18 @@ export class EventManager {
+export abstract class EventManagerPlugin {
+    abstract addEventListener(element: HTMLElement, eventName: string, handler: (event: Event) => void): Function;
+    // (undocumented)
+    manager: EventManager;
```
Great idea, that property bothered me also !

waterplea replied:
Without that property, the example doesn't work, as well as the broader concept of unintrusive plugins that do their part and delegate further handling back to the manager which is heavily used by [@tinkoff/ng-event-plugins](https://github.com/Tinkoff/ng-event-plugins).

JeanMeche replied:
The problem of accessing the manager can be solved by injecting it ! 

```typescript
export class StopEventPlugin {
  constructor(private injector: Injector) {}

  supports(event: string): boolean {
    return event.endsWith('.stop');
  }

  addEventListener(element: HTMLElement, eventName: string, handler: (event: Event) => void): Function {
    const eventWrapper = (event: Event) => {
      event.stopPropagation();
      handler(event);
    };

    // striping the stop modifier from the event name;
    const nonModifiedEventName = eventName.replace(/\.stop$/, '');

    const manager = this.injector.get(EventManager);
    return manager.addEventListener(element, nonModifiedEventName, eventWrapper);
  }
}
```

What do you think ?  Should we revert the breaking change or suggest to inject the manager if it's needed ?

waterplea replied:
Personally, I don't think switching from property assignment by `EventManager` to the entire `Injector` to avoid cyclic dependency is much better. Sounds like for plugins to have direct access to the manager is reasonable.

atscott replied:
> Sounds like for plugins to have direct access to the manager is reasonable.

Having access to the manager isn't really the issue, but rather the awkwardness of having a public API where the some other service assigns a value to a user's publicly writable property after the service is created.

waterplea replied:
Maybe use the same approach control value accessors have and add `registerManager(manager: EventManager): void`?

atscott commented:
@JeanMeche Can you separate this in to two PRs? One that simply adds the `@publicApi` annotation to `EventManagerPlugin` and another that adds the documentation? For the public API change, you can simply cite the existing token: https://angular.io/api/platform-browser/EVENT_MANAGER_PLUGINS. The exposed type of the token should definitely be in the public API.

JeanMeche commented:
@atscott I've reduced the scope of the PR to adding `EventManagerPlugin` to the public API.

Do you have any clue why `getFactoryOf` has been added to the goldens ?

atscott commented:
> @atscott I've reduced the scope of the PR to adding EventManagerPlugin to the public API.

> Do you have any clue why getFactoryOf has been added to the goldens ?

The PR is still a bit more than adding `EventManagerPlugin` to the public API. There are some other refactors in here that caused the golden files to change and `getFactoryOf` is a result of those refactors. I believe they are there now because you inherit the constructor from the base class rather than having the explicit `super` calls. Could you limit the changes even further?

JeanMeche commented:
@atscott I wanted to clean a bit the API before making it public, as its always hard to update once it's public.  

* `_doc` in `EventManaerPlugin` is unused so I removed it. 
* I update all `super()` calls as a result. 
* `Function` type for the handler is really vague and I wanted to improve it. 

By restoring the constructor, I got to revert the changes in the golden. 

What do you think of it now ?

atscott reviewed (approved):
> @atscott I wanted to clean up a bit the API before making it public, as its always harder to update once it's public.

Yea, that's a good idea. The changes generally look good to me now.

Note: While abstract, `EventManagerPlugin` is still technically not allowed to be extended by developers due to our "classes are final unless explicitly noted otherwise" policy. You will need to add that developers can extend the class if that's the intent (similar to https://github.com/angular/angular/blob/e40a640dfe54b03bfe917d08098c319b0b200d25/packages/router/src/route_reuse_strategy.ts#L76-L77)

JeanMeche commented on packages/platform-browser/src/dom/events/key_events.ts:
```diff
@@ -45,17 +45,12 @@ const MODIFIER_KEY_GETTERS: {[key: string]: (event: KeyboardEvent) => boolean} =
 
 /**
- * @publicApi
  * A browser plug-in that provides support for handling of key events in Angular.
```
It wasn't public, it's exported as `ɵKeyEventsPlugin`.

alfaproject commented on packages/platform-browser/src/dom/events/dom_events.ts:
```diff
@@ -6,15 +6,14 @@
 export class DomEventsPlugin extends EventManagerPlugin {
-  constructor(@Inject(DOCUMENT) doc: any) {
-    super(doc);
+  constructor() {
```
These should be redundant now

JeanMeche replied:
Actually they aren't redundant. 

Having no constructor pulls `getFactoryOf()` cf my comment above about the goldens & Andrew's answer.

alfaproject replied:
What kind of sorcery is that? o:

JeanMeche replied:
I believe its down to the implementation of `ɵɵgetInheritedFactory` : 
https://github.com/angular/angular/blob/8a2739f25034443f3a72d57f983e70c1a9d69722/packages/core/src/render3/di.ts#L722-L751

It calls getFactoryOf() if there is no constructor on the class itself.

alfaproject commented on packages/platform-browser/src/dom/events/key_events.ts:
```diff
@@ -45,17 +45,12 @@ const MODIFIER_KEY_GETTERS: {[key: string]: (event: KeyboardEvent) => boolean} =
-   */
-  constructor(@Inject(DOCUMENT) doc: any) {
-    super(doc);
+  constructor() {
```
Redundant now, right?

alfaproject commented on packages/platform-browser/src/dom/events/key_events.ts:
```diff
@@ -188,7 +183,6 @@ export class KeyEventsPlugin extends EventManagerPlugin {
 
   /** @internal */
   static _normalizeKey(keyName: string): string {
-    // TODO: switch to a Map if the mapping grows too much
```
Removing this comment is unrelated to this PR. You probably should revert and actually refactor this to a simple if condition until a switch or map or whatever else is needed because an if condition minifies better

JeanMeche replied:
Yeah good idea 👍

jessicajaniuk reviewed (approved):
reviewed-for: public-api

This will definitely need a TGP.

dylhunn reviewed (approved):
reviewed-for: public-api

josephperrott commented:
@dylhunn replying offline as this is a g3 specific item.

dylhunn commented:
I will run a TGP on [this CL](https://critique.corp.google.com/cl/530642500) tonight.

JeanMeche commented:
Any news @dylhunn ?

JeanMeche committed:
feat(platform-browser): expose `EventManagerPlugin` in the public API.

    The exposed type of the `EVENT_MANAGER_PLUGINS` token should be in the public API.

dylhunn commented:
TGP is running: https://test.corp.google.com/OCL:560756113:BASE:560894821:1693279799412:6c3fde97

dylhunn commented:
TGP passed.

jessicajaniuk commented:
This PR was merged into the repository by commit c5daa6ce776724d44c02cc97f1a349a85cb2a819.
